File: rs2-core/src/services/query.rs
//! `query` — stored parameterized queries (PRD §10.4), store-patterned.
//!
//! Authoring lives under the reserved subtree `/<mount>/.queries/…` — a
//! normal store-contract surface delegated to the owned `FileService` via
//! [`SpecStore`] (validation at write time). Every other path, on **any
//! verb**, executes: the longest stored prefix wins (peeled segments become
//! positional params `"0"`, `"1"`, …; a `.root` spec governs the mount
//! root), so a stored query can serve plain `GET`.
//!
//! Execution parameters, later sources winning: positional URL segments →
//! query-string pairs (coerced to the params schema's declared types) →
//! JSON body (object named / array positional). Defaults apply, then the
//! whole set validates against the envelope's `params` schema.

use async_trait::async_trait;
use serde_json::{Map, Value};

use super::query_template::{self, Language, QueryEnvelope};
use super::spec_store::SpecStore;
use super::{pagination, Service, ServiceContext};
use crate::error::RsError;
use crate::message::Message;

/// Default storage prefix (under the tenant file store) for query mounts.
pub const QUERY_PREFIX: &str = ".rs2-queries";

/// The reserved authoring subtree segment.
pub const QUERY_SUBTREE: &str = ".queries";

pub struct QueryService {
    store: SpecStore,
}

impl QueryService {
    pub fn from_config(config: &Value, store: SpecStore) -> Result<Self, RsError> {
        if config.get("queries").is_some() {
            return Err(RsError::bad_request(
                "config-defined queries are no longer supported: PUT query envelopes to \
                 /<mount>/.queries/<name> instead",
            ));
        }
        Ok(QueryService { store })
    }

    /// The write-time validator: the envelope must parse (schemas compile,
    /// placeholders scan); stored as submitted.
    pub fn validator() -> super::spec_store::SpecValidator {
        std::sync::Arc::new(|doc: &Value| {
            QueryEnvelope::parse(doc)?;
            Ok(doc.clone())
        })
    }
}

#[async_trait]
impl Service for QueryService {
    async fn handle(&self, mut msg: Message, ctx: &ServiceContext) -> Result<Message, RsError> {
        if self.store.is_authoring(&msg) {
            return self.store.handle_authoring(msg).await;
        }

        // ---- execution: any verb, longest stored prefix ----
        let store = ctx
            .query
            .as_ref()
            .ok_or_else(|| RsError::internal("query service has no QueryStore capability"))?;
        let segments: Vec<String> = msg
            .url
            .service_segments()
            .iter()
            .map(|s| s.to_string())
            .collect();
        let Some((doc, split)) = self.store.resolve(&segments).await? else {
            return Err(RsError::not_found(format!(
                "no stored query matches '{}' (author one at {}{}/…)",
                msg.url.service_path, msg.url.base_path, QUERY_SUBTREE,
            )));
        };
        let envelope = QueryEnvelope::parse(&doc)?;

        // Parameters: positional URL segments, then query-string pairs
        // (schema-coerced), then the JSON body — later wins.
        let mut params = Map::new();
        for (i, seg) in segments[split..].iter().enumerate() {
            params.insert(i.to_string(), Value::String(seg.clone()));
        }
        params.extend(query_template::url_params(
            &msg.url.query,
            envelope.params_schema.as_ref(),
        ));
        match &mut msg.body {
            None => {}
            Some(b) => match b.as_json(ctx.limits.materialized_body_bytes).await? {
                Value::Object(named) => params.extend(named),
                Value::Array(items) => {
                    for (i, v) in items.into_iter().enumerate() {
                        params.insert(i.to_string(), v);
                    }
                }
                Value::Null => {}
                _ => {
                    return Err(RsError::bad_request(
                        "query parameters must be a JSON object or array",
                    ))
                }
            },
        }
        let params = envelope.prepare_params(params)?;

        // JSON templates substitute structurally here; string templates
        // (SQL) pass through for the adapter to bind.
        let query = match envelope.language {
            Language::Json => {
                let quote = |v: &Value| store.quote(v);
                query_template::substitute_json(&envelope.query, &params, &quote)?
            }
            Language::Text => envelope.query.clone(),
        };

        let (take, skip) = pagination(&msg);
        let (rows, total) = store.run_query(&query, &params, take, skip).await?;

        let mut resp = msg.ok_json(&Value::Array(rows));
        resp.set_header("x-total-count", &total.to_string());
        Ok(resp)
    }
}

File: rs2-worker/src/services/query.ts
// `query` — stored parameterized queries (PRD §10.4), store-patterned.
// Authoring lives under the reserved subtree `/<mount>/.queries/…` — a
// normal store-contract surface delegated to the owned FileService via the
// SpecStore (validation at write time). Every other path, on **any verb**,
// executes: the longest stored prefix wins (peeled segments become
// positional params `"0"`, `"1"`, …; a `.root` spec governs the mount
// root), so a stored query can serve plain `GET`. Port of
// `rs2-core/src/services/query.rs`.
//
// Execution parameters, later sources winning: positional URL segments →
// query-string pairs (coerced to the params schema's declared types) →
// JSON body (object named / array positional). Defaults apply, then the
// whole set validates against the envelope's `params` schema.

import { RsError } from "../runtime/error";
import type { Json, JsonObject } from "../runtime/error";
import type { Message } from "../runtime/message";
import { QUERY_SUBTREE, pagination } from "./context";
import type { Service, ServiceContext } from "./context";
import { QueryEnvelope, substituteJson, urlParams } from "./query-template";
import type { SpecStore, SpecValidator } from "./spec-store";

export class QueryService implements Service {
  private constructor(private readonly store: SpecStore) {}

  static fromConfig(config: JsonObject, store: SpecStore): QueryService {
    if (config.queries !== undefined) {
      throw RsError.badRequest(
        "config-defined queries are no longer supported: PUT query envelopes to /<mount>/.queries/<name> instead",
      );
    }
    return new QueryService(store);
  }

  /// The write-time validator: the envelope must parse (schemas compile,
  /// placeholders scan); stored as submitted.
  static validator(): SpecValidator {
    return (doc: Json): Json => {
      QueryEnvelope.parse(doc);
      return doc;
    };
  }

  async handle(msg: Message, ctx: ServiceContext): Promise<Message> {
    if (this.store.isAuthoring(msg)) return this.store.handleAuthoring(msg);

    // ---- execution: any verb, longest stored prefix ----
    const store = ctx.query;
    if (!store) throw RsError.internal("query service has no QueryStore capability");
    const segments = msg.url.serviceSegments();
    const resolved = await this.store.resolve(segments);
    if (!resolved) {
      throw RsError.notFound(
        `no stored query matches '${msg.url.servicePath}' (author one at ${msg.url.basePath}${QUERY_SUBTREE}/…)`,
      );
    }
    const [doc, split] = resolved;
    const envelope = QueryEnvelope.parse(doc);

    // Parameters: positional URL segments, then query-string pairs
    // (schema-coerced), then the JSON body — later wins.
    const params: JsonObject = {};
    segments.slice(split).forEach((seg, i) => {
      params[String(i)] = seg;
    });
    Object.assign(params, urlParams(msg.url.query, envelope.paramsSchema));
    if (msg.body) {
      const body = await msg.body.asJson(ctx.limits.materializedBodyBytes);
      if (Array.isArray(body)) {
        body.forEach((v, i) => {
          params[String(i)] = v;
        });
      } else if (body && typeof body === "object") {
        Object.assign(params, body);
      } else if (body !== null) {
        throw RsError.badRequest("query parameters must be a JSON object or array");
      }
    }
    const prepared = envelope.prepareParams(params);

    // JSON templates substitute structurally here; string templates (SQL)
    // pass through for the adapter to bind.
    const query =
      envelope.language === "json" ? substituteJson(envelope.query, prepared, (v) => store.quote(v)) : envelope.query;

    const [take, skip] = pagination(msg);
    const [rows, total] = await store.runQuery(query, prepared, take, skip);

    const resp = msg.okJson(rows);
    resp.setHeader("x-total-count", String(total));
    return resp;
  }
}
